src/algebraic_geometry/EllipticCurve/point.lean

```diff
@@ -0,0 +1,447 @@
+                    + (-↑E.a₁ * x₁ ^ 2 - 3 * ↑E.a₁ * x₁ * x₂ - 4 * x₁ * y₁ - 2 * ↑E.a₁ * x₂ ^ 2
+                        - 2 * x₂ * y₁ - ↑E.a₁ * ↑E.a₂ * x₁ - 2 * ↑E.a₃ * x₁ - 2 * ↑E.a₁ * ↑E.a₂ * x₂
+                        - ↑E.a₃ * x₂ + ↑E.a₁ ^ 2 * y₁ - 2 * ↑E.a₂ * y₁ - ↑E.a₁ * ↑E.a₄
+                        - ↑E.a₂ * ↑E.a₃))
+    + (x₁ ^ 3 + 3 * x₁ ^ 2 * x₂ + 3 * x₁ * x₂ ^ 2 + x₂ ^ 3 + 2 * ↑E.a₂ * x₁ ^ 2
+        + 4 * ↑E.a₂ * x₁ * x₂ - ↑E.a₁ * x₁ * y₁ + 2 * ↑E.a₂ * x₂ ^ 2 - ↑E.a₁ * x₂ * y₁ + y₁ ^ 2
+        + ↑E.a₂ ^ 2 * x₁ + ↑E.a₄ * x₁ + ↑E.a₂ ^ 2 * x₂ + ↑E.a₄ * x₂ - ↑E.a₁ * ↑E.a₂ * y₁
+        + ↑E.a₃ * y₁ + ↑E.a₂ * ↑E.a₄ - ↑E.a₆) = 0,
+  { simp only [add_x_def, add_y'_def], ring1 },
+  field_simp [sub_ne_zero_of_ne hx],
+  rw [weierstrass_equation_iff'] at h₁ h₂,
+  linear_combination
+    -((x₁ - x₂) ^ 2 * (x₁ + 2 * x₂ + ↑E.a₂) - (x₁ - x₂) * (y₁ - y₂) * ↑E.a₁ - (y₁ - y₂) ^ 2) * h₁
+    + ((x₁ - x₂) ^ 2 * (2 * x₁ + x₂ + ↑E.a₂) - (x₁ - x₂) * (y₁ - y₂) * ↑E.a₁ - (y₁ - y₂) ^ 2) * h₂
+    with { normalization_tactic := `[ring1] }
+end
+
+/-- The addition of two affine points in `E` with distinct $x$-coordinates lies in `E`. -/
+lemma weierstrass_equation_add (hx : x₁ ≠ x₂) :
+  E.weierstrass_equation (add_x h₁ h₂) (add_y h₁ h₂) :=
+weierstrass_equation_neg $ weierstrass_equation_add' h₁ h₂ hx
+
+lemma y_eq_of_y_ne (hx : x₁ = x₂) (hy : y₁ ≠ neg_y h₂) : y₁ = y₂ :=
+begin
+  rw [neg_y] at hy,
+  rw [weierstrass_equation_iff] at h₂,
+  rw [weierstrass_equation_iff', hx, ← h₂] at h₁,
+  apply eq_of_sub_eq_zero ∘ eq_zero_of_ne_zero_of_mul_right_eq_zero (sub_ne_zero.mpr hy),
+  convert h₁,
+  ring1
+end
+
+lemma y_ne_of_y_ne (hx : x₁ = x₂) (hy : y₁ ≠ neg_y h₂) : y₁ ≠ neg_y h₁ :=
+by { convert hy, exact y_eq_of_y_ne h₁ h₂ hx hy }
+
+omit h₁ h₂
+
+open_locale classical
+
+/-- Addition in `E⟮K⟯`. Given `P Q : E⟮K⟯`, use `P + Q` instead of `add P Q`. -/
+noncomputable def add : E⟮K⟯ → E⟮K⟯ → E⟮K⟯
+| 0               P               := P
+| P               0               := P
+| (some x₁ y₁ h₁) (some x₂ y₂ h₂) :=
+if hx : x₁ = x₂ then if hy : y₁ = neg_y h₂ then 0
+else some _ _ $ weierstrass_equation_dbl h₁ $ y_ne_of_y_ne h₁ h₂ hx hy
+else some _ _ $ weierstrass_equation_add h₁ h₂ hx
+
+noncomputable instance : has_add E⟮K⟯ := ⟨add⟩
+
+@[simp] lemma add_def (P Q : E⟮K⟯) : add P Q = P + Q := rfl
+
+@[simp] lemma zero_add (P : E⟮K⟯) : 0 + P = P := by cases P; refl
+
+@[simp] lemma add_zero (P : E⟮K⟯) : P + 0 = P := by cases P; refl
+
+noncomputable instance : add_zero_class E⟮K⟯ := ⟨0, (+), zero_add, add_zero⟩
+
+lemma some_add_some :
+  some x₁ y₁ h₁ + some x₂ y₂ h₂
+    = if hx : x₁ = x₂ then if hy : y₁ = neg_y h₂ then 0
+      else some _ _ $ weierstrass_equation_dbl h₁ $ y_ne_of_y_ne h₁ h₂ hx hy
+      else some _ _ $ weierstrass_equation_add h₁ h₂ hx :=
+rfl
+
+@[simp] lemma some_add_some_of_y_eq (hx : x₁ = x₂) (hy : y₁ = neg_y h₂) :
+  some x₁ y₁ h₁ + some x₂ y₂ h₂ = 0 :=
+by rw [some_add_some, dif_pos hx, dif_pos hy]
```

Comment:
Once you have `add_zero_class` then the simp lemmas [zero_add](https://leanprover-community.github.io/mathlib_docs/algebra/group/defs.html#zero_add) automatically apply. Also `rw some_add_some` can be replaced by `rw [← add_def, add]`, so:
```suggestion
noncomputable instance : add_zero_class E⟮K⟯ :=
⟨0, (+), by rintro (P|P); refl, by rintro (P|P); refl⟩

@[simp] lemma some_add_some_of_y_eq (hx : x₁ = x₂) (hy : y₁ = neg_y h₂) :
  some x₁ y₁ h₁ + some x₂ y₂ h₂ = 0 :=
by rw [← add_def, add, dif_pos hx, dif_pos hy]
```